src/api/v1/db/oracledb/pizzas-dao.js

```diff
@@ -0,0 +1,112 @@
+import _ from 'lodash';
+
+import { getConnection } from 'api/v1/db/oracledb/connection';
+import { getDoughById } from 'api/v1/db/oracledb/doughs-dao';
+import { serializePizza } from 'api/v1/serializers/pizzas-serializer';
+
+const pizzaColumns = {
+  id: 'ID',
+  doughId: 'DOUGH_ID',
+  name: 'NAME',
+  bakeTime: 'BAKE_TIME',
+  ovenTemp: 'OVEN_TEMP',
+  specialInstructions: 'SPECIAL_INSTRUCTIONS',
+};
+
+/**
+ * A list of columns in the PIZZAS table aliased to their
+ * corresponding PizzaResource attribute names
+ *
+ * @const {string}
+ */
+const pizzaColumnAliases = _.map(pizzaColumns,
+  (columnName, attributeName) => `${columnName} AS "${attributeName}"`)
+  .join(', ');
+
+/**
+ * A SQL query to get a pizza with a specific ID
+ *
+ * @const {string}
+ */
+const getPizzaByIdQuery = `SELECT ${pizzaColumnAliases} FROM PIZZAS WHERE ID = :id`;
+
+/**
+ * Fetch all the ingredients associated with the pizza with id
+ *  `pizzaId`
+ *
+ * @param {string} pizzaId
+ * @returns {Array} an array of unserialized ingredient objects
+ */
+const getPizzaIngredients = async (pizzaId) => {
+  const query = `SELECT
+    INGREDIENTS.ID AS "id",
+      INGREDIENTS.TYPE AS "ingredientType",
+      INGREDIENTS.NAME AS "name",
+      INGREDIENTS.NOTES AS "notes"
+    FROM PIZZA_INGREDIENTS INNER JOIN INGREDIENTS ON
+      PIZZA_INGREDIENTS.INGREDIENT_ID = INGREDIENTS.ID
+    WHERE PIZZA_INGREDIENTS.PIZZA_ID = :id`;
+  const bindParams = { id: pizzaId };
+  const connection = await getConnection();
+  try {
+    const result = await connection.execute(
+      query,
+      bindParams,
+    );
+    return result.rows;
+  } finally {
+    connection.close();
+  }
+};
+
+/**
+ * Get a "deserialized" dough object using `getDoughById`
+ *
+ * @param {string} doughId the ID of the dough to get
+ * @returns {object} a raw dough object
+ */
+const getPizzaDough = async (doughId) => {
+  const serializedDough = await getDoughById(doughId);
+
+  if (serializedDough === null) return {};
+
+  const deserializedDough = {
+    id: serializedDough.data.id,
+    ...serializedDough.data.attributes,
+  };
+  return deserializedDough;
+};
+
+/**
+ * Fetch the pizza with ID `pizzaId`
+ *
+ * @param {string} pizzaId
+ * @returns {Promise<object>} the full pizza record
+ */
+const getPizzaById = async (pizzaId) => {
+  const bindParams = { id: pizzaId };
+  const connection = await getConnection();
+  try {
+    const { rows } = await connection.execute(getPizzaByIdQuery, bindParams);
+
+    if (rows.length > 1) throw new Error('Returned multiple outputs for the same ID');
+    if (rows.length === 0) return null;
+
+    const { doughId } = rows[0];
+
+    const dough = await getPizzaDough(doughId);
+    const ingredients = await getPizzaIngredients(pizzaId);
```

Comment:
It works, but you actually sent three SQL calls to get one pizza. It's not a big deal right now, but it will be a serious issue when you implement `GET /pizzas` endpoint. Can you try combining them into one SQL query?

Reply:
I was concerned about the performance impacts having to join across four tables when fetching all of the pizza resources, so I was actually planning on asking about this. Do you think it's basically guaranteed that even one really expensive query is better than making a number of smaller queries? 

I'm also not sure what optimization strategies we would have available for either option. I had the thought that you could potentially cache the smaller queries on the server side, but I don't really know about what you can do to improve performance on the database side.

Reply:
Really a great question, I also chated with Jon (our DBA) about this topic.
To be honest, it really depends on how you destructure your query, but in this case and most of our cases, larger query will still be more efficient than multiple smaller queries. Caching can improve performance for sure, but you need to make sure the smaller queries used frequently enough so that they can be cached, otherwise it will only increase the cost.

Reply:
Good point about caching. Going forward I'll switch to using a single query; I'm going to implement `includes` first using the current code and then I'll get the query updated.